cmd/service.go

```diff
@@ -65,3 +67,57 @@ func restartProcess() error {
 	// Re-uses the same pid. This preserves the pid over multiple server-respawns.
 	return syscall.Exec(argv0, os.Args, os.Environ())
 }
+
+// Keep track of number of freeze/unfreeze calls.
+const trackFreezeCount = true
+
+// freezeServices will freeze all incoming S3 API calls.
+// For each call, unfreezeServices must be called once.
+func freezeServices() {
+	// Use atomics for globalServiceFreeze, so we can read without locking.
+	if trackFreezeCount {
+		// We need a lock since we are need the 2 atomic values to remain in sync.
+		globalServiceFreezeMu.Lock()
+		// If multiple calls, first one creates channel.
+		globalServiceFreezeCnt++
+		if globalServiceFreezeCnt == 1 {
+			globalServiceFreeze.Store(make(chan struct{}))
+		}
+		globalServiceFreezeMu.Unlock()
+	} else {
+		// If multiple calls, first one creates channel.
+		globalServiceFreeze.CompareAndSwap(nil, make(chan struct{}))
```

Comment:
> `func (*Value) CompareAndSwap <-- added in go1.17`

Ah, too new. This can be 
```suggestion
		globalServiceFreeze.Store(nil, make(chan struct{}))
```
... but in `unfreezeServices` it is a bit trickier without the lock.

Reply:
yeah well, we can take out go1.16 @klauspost